File: src/lowlevel/client/icefsoperators/icefsReleaseDir.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:41:00
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsReleaseDir.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsReleaseDir(fuse_req_t fuseReq, fuse_ino_t inode,
                                    struct fuse_file_info *fi) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsReleaseDirReq req;
      icefsgrpc::IcefsReleaseDirRes res;
      grpc::ClientContext ctx;

      req.set_fh(fi->fh);

      grpc::Status status = gRpcClient->DoIcefsReleaseDir(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_err(fuseReq, ICEFS_EOK);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsReleaseDirReq req;
      icefsthrift::IcefsReleaseDirRes res;

      req.__set_fh(fi->fh);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsReleaseDir(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      fuse_reply_err(fuseReq, res.status);
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}

File: src/lowlevel/server/icefsoperators/icefsReleaseDir.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 13:45:24
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsReleaseDir.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
)

func (s *IcefsServer) doIcefsReleaseDir(fh uint64) (status int32) {
	s.delIcefsDir(fh)
	return int32(icefserror.ICEFS_EOK)
}

func (s *IcefsGRpcServer) DoIcefsReleaseDir(ctx context.Context, req *pb.IcefsReleaseDirReq) (*pb.IcefsReleaseDirRes, error) {
	var res pb.IcefsReleaseDirRes

	res.Status = s.server.doIcefsReleaseDir(req.Fh)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsReleaseDir(ctx context.Context, req *icefsthrift.IcefsReleaseDirReq) (*icefsthrift.IcefsReleaseDirRes, error) {
	var res icefsthrift.IcefsReleaseDirRes

	res.Status = s.server.doIcefsReleaseDir(uint64(req.Fh))

	return &res, nil
}
